src/IR/Simplify.hs

```diff
@@ -0,0 +1,459 @@
+          pure (Nothing, sub')
+        _ -> do
+          -- preinline test FAILS, so do post inline unconditionally
+          e' <- simplExpr sub ins rhs cont -- process the RHS
+          case e' of
+            -- x goes here.
+            (I.Lit _ _) -> do
+              insertSubst v $ DoneEx e'
+              pure (Nothing, M.empty) -- PASSES postinline
+            (I.Var _ _) -> do
+              insertSubst v $ DoneEx e'
+              pure (Nothing, M.empty) -- PASSES postinline
+            _ -> do
+              rhs' <- simplExpr sub ins rhs cont -- we won't inline x, but still possible to simplify e (RHS)
+              pure (Just (binder, rhs'), sub) -- FAILS postinline; someday callsite inline
+      _ -> do
+        e' <- simplExpr sub ins rhs cont
+        pure (Just (binder, e'), sub) -- can't inline wildcards
+
+-- for all other expressions, don't do anything
+simplExpr _ _ e _ = pure e
+
+
+{- | Run occurrence analyser over each top level function
+
+Returns logging info as a string.
+-}
+runOccAnal :: I.Program I.Type -> SimplFn String
+runOccAnal I.Program{I.programDefs = defs} = do
+  -- a hacky way to see the occurence info for each top def
+  let getOccInfoForDef ::
+        (I.VarId, SimplFn (I.Expr I.Type, String)) -> SimplFn String
+      getOccInfoForDef (v, tpl) = do
+        (_, occinfo) <- tpl
+        pure $
+          "START topdef "
+            ++ show v
+            ++ " has OccInfo: "
+            ++ show occinfo
+            ++ " END"
+  defs' <- mapM swallowArgs defs
+  info <- mapM (getOccInfoForDef . second occAnalExpr) defs'
+  return (show info)
+ where
+  {- Take in a top level function, "swallows" its args, and return its body.
+
+  "Swallow" means to add the argument to our occurrence info state.
+  It returns a top level function without curried arguments; just the body.
+  -}
+  swallowArgs :: (I.VarId, I.Expr t) -> SimplFn (I.VarId, I.Expr t)
+  swallowArgs (funcName, l@(I.Lambda _ _ _)) = do
+    addOccs (Just funcName) -- HACKY
+    let (args, body) = unfoldLambda l
+    mapM_ addOccs args
+    pure (funcName, body)
+   where
+    addOccs (Just nm) = addOccVar nm
+    addOccs Nothing = pure ()
+  swallowArgs (name, e) = pure (name, e)
+
+
+{- | Run the Ocurrence Analyzer over an IR expression node
+
+How do we handle each node?
+- Var VarId t                           DONE
+- Data DConId t                         Default case (TODO: trivial constructor argument invariant)
+- Lit Literal t                         Default case
+- App (Expr t) (Expr t) t               DONE
+- Let [(Binder, Expr t)] (Expr t) t     DONE
+- Lambda Binder (Expr t) t              DONE
+- Match (Expr t) [(Alt, Expr t)] t      DONE (TODO: analyze patterns / LHS of arms?)
+- Prim Primitive [Expr t] t             DONE
+-}
+occAnalExpr :: I.Expr I.Type -> SimplFn (I.Expr I.Type, String)
+-- Occurrence Analysis over Let Expression
+occAnalExpr l@(I.Let nameValPairs body _) = do
+  mapM_
```

Comment:
You can use `forM_` here, which is just defined as follows:

```haskell
forM_ = flip mapM_
```

Then you don't need to wrap the lambda around parens, you can just do

```haskell
forM_ nameValPairs $ \(binder, rhs) -> do $ -- etc.
```

Reply:
(you'll also need to import `forM` from the standard library)